### PR TITLE
Add support for Windows builds with the Ninja generator

### PR DESCRIPTION
**Pull request by thewtex**

### CODE REVIEW

**Comment by thewtex** [comment]
@jessdtate @allywarner @jcfr please take a look

**Review by jessdtate — DISMISSED** [review]

**Comment by jcfr** [comment]
After a quick look, using CMAKE_BUILD_TYPE on windows does not seem right. I will have another look tomorrow.

**Review by allywarner — DISMISSED** [review]

**Comment by allywarner** [comment]
Waiting on @jcfr approval before merging.

File: src/test/CMakeLists.txt
Comment (jcfr) [inline src/test/CMakeLists.txt:18]: Checking for `WIN32` and then comparing `CMAKE_BUILD_TYPE` does not seem valid. This should be revisited to check `CMAKE_CONFIGURATION_TYPES` and/or using generator expression

Comment (jcfr) [reply src/test/CMakeLists.txt:18]: Similarly to what was done below

**Review by jcfr — CHANGES_REQUESTED** [review]
Few comments

Consider also rebasing against master instead of merging it into this topic

File: src/test/CMakeLists.txt
Comment (thewtex) [reply src/test/CMakeLists.txt:38]: Done

File: src/test/CMakeLists.txt
Comment (thewtex) [reply src/test/CMakeLists.txt:18]: While I also tried a generator expression initially `BUILD_BYPRODUCTS` does not support generator expressions. For the *Ninja* generator, we use `CMAKE_BUILD_TYPE`.  I tested this patch with both the *Ninja* generator and the *Visual Studio* generator.

Comment (jcfr) [reply src/test/CMakeLists.txt:18]: Make sense, I suggest we do the conditional on the generator then.

Comment (jcfr) [reply src/test/CMakeLists.txt:18]: Because the current implementation assumes that `CMAKE_BUILD_TYPE` is defined on `WIN32` and this is not always the case.

Comment (jcfr) [reply src/test/CMakeLists.txt:18]: I suggest the following:
* introduce `_ep_arg_build_byproducts` internal var
* set it to `_ep_arg_build_byproducts BUILD_BYPRODUCTS "${_build_byproducts}"` only if the generator is Ninja or if a single configuration generator

Comment (thewtex) [reply src/test/CMakeLists.txt:18]: Done

File: src/test/CMakeLists.txt
Comment (jcfr) [inline src/test/CMakeLists.txt:34]:
```suggestions
set(_cmake_cache_args)
```

File: src/test/CMakeLists.txt
Comment (jcfr) [inline src/test/CMakeLists.txt:36]:
```suggestion
  list(APPEND _cmake_cache_args -DCMAKE_BUILD_TYPE:STRING=${CMAKE_BUILD_TYPE})
```

File: src/test/CMakeLists.txt
Comment (jcfr) [inline src/test/CMakeLists.txt:43]:
```suggestion
     ${_cmake_cache_args}
```

File: src/test/CMakeLists.txt
Comment (jcfr) [inline src/test/CMakeLists.txt:119]:
```suggestion
    else()
      # Ninja
      target_link_libraries(${name}
       ${binary_dir}/lib/${CMAKE_FIND_LIBRARY_PREFIXES}gtest$<$<CONFIG:Debug>:d>${CMAKE_STATIC_LIBRARY_SUFFIX}
       ${binary_dir}/lib/${CMAKE_FIND_LIBRARY_PREFIXES}gtest_main$<$<CONFIG:Debug>:d>${CMAKE_STATIC_LIBRARY_SUFFIX}
       ${binary_dir}/lib/${CMAKE_FIND_LIBRARY_PREFIXES}gmock$<$<CONFIG:Debug>:d>${CMAKE_STATIC_LIBRARY_SUFFIX}
       ${binary_dir}/lib/${CMAKE_FIND_LIBRARY_PREFIXES}gmock_main$<$<CONFIG:Debug>:d>${CMAKE_STATIC_LIBRARY_SUFFIX}
        cleaver )
    endif()
```

**Review by jcfr — APPROVED** [review]
Consider doing "squash + merge" adding the following commit message:

```
COMP: Add support for Windows builds with the Ninja generator

This commit is a follow up of 0f0137a27 (COMP: Add support for
Ninja CMake Generator)

Co-authored-by: Matt McCormick <matt.mccormick@kitware.com>
Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>
```

File: src/test/CMakeLists.txt
Comment (thewtex) [reply src/test/CMakeLists.txt:34]: Done

File: src/test/CMakeLists.txt
Comment (thewtex) [reply src/test/CMakeLists.txt:119]: Done

**Comment by thewtex** [comment]
Rebase opened in https://github.com/SCIInstitute/Cleaver2/pull/109

**Pull request closed without merging**
